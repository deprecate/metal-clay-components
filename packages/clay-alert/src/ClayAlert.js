import 'clay-button';
import 'clay-icon';
import Component from 'metal-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayAlert.soy.js';

/**
 * Metal Clay Alert component.
 */
class ClayAlert extends Component {
  /**
	 * Handles onclick event for the close button in case of closeable alert.
	 * @param {!Event} event
	 * @private
	 */
  _handleCloseClick(event) {
    this.emit('close');
  }
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayAlert.STATE = {
  /**
	 * If the alert is closeable
	 * @instance
	 * @memberof ClayAlert
	 * @type {?bool}
	 * @default false
	 */
  closeable: Config.bool().value(false),

  /**
	 * The message of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {!string}
	 */
  message: Config.string().required(),

  /**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayIcon
	 * @type {?string}
	 * @default undefined
	 */
  spritemap: Config.string(),

  /**
	 * The style of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {?string}
	 * @default info
	 */
  style: Config.oneOf(['danger', 'info', 'success', 'warning']).value('info'),

  /**
	 * The type of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {?string}
	 * @default embedded
	 */
  type: Config.oneOf(['embedded', 'stripe', 'toast']).value('embedded'),

  /**
	 * The title of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {!string}
	 */
  title: Config.string().required(),
};

Soy.register(ClayAlert, templates);

export {ClayAlert};
export default ClayAlert;
