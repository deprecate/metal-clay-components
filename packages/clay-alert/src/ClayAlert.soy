{namespace ClayAlert}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param message: html}
	{@param spritemap: string}
	{@param title: string}
	{@param? _handleCloseClick: any}
	{@param? _visible: bool}
	{@param? closeable: bool}
	{@param? style: string}
	{@param? type: string}

	{let $attributes kind="attributes"}
		class="alert alert-{$style} fade
			{if $closeable}
				{sp}alert-dismissible
			{/if}

			{if $type}
				{sp}alert-{$type}
			{/if}

			{if $_visible != false}
				{sp}show
			{/if}
		"
		role="alert"
	{/let}

	<div {$attributes}>
		{if $closeable}
			{call ClayButton.render}
				{param ariaLabel: 'Close' /}
				{param elementClasses: 'close' /}
				{param events: [ 'click': $_handleCloseClick ] /}
				{param icon: [
					'alignment': 'right',
					'monospaced': false,
					'spritemap': $spritemap,
					'symbol': 'times'
					] /}
				{param style: 'unstyled' /}
				{param type: 'button' /}
			{/call}
		{/if}

		{let $icon kind="text"}
			{switch $style}
				{case 'danger'}
					exclamation-full
				{case 'success'}
					check-circle
				{case 'warning'}
					question-circle-full
				{default}
					info-circle
			{/switch}
		{/let}

		{call ClayIcon.render}
			{param spritemap: $spritemap /}
			{param symbol: $icon /}
		{/call}

		<strong class="lead">
			{$title}
		</strong>

		{$message}
	</div>
{/template}