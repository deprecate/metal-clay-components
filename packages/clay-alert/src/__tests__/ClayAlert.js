import ClayAlert from '../ClayAlert';

let alert;

const spritemap = '../node_modules/lexicon-ux/build/images/icons/icons.svg';

describe('ClayAlert', function() {
	afterEach(() => {
		if (alert) {
			alert.dispose();
		}
	});

	it('should render default markup', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a notification alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
			type: 'notification',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a stripe alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
			type: 'stripe',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert style "info"', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert style "danger"', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'danger',
			title: 'danger',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert style "warning"', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'warning',
			title: 'warning',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert style "success"', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'success',
			title: 'success',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a closeable alert', function() {
		alert = new ClayAlert({
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			title: 'success',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should close an alert', function() {
		alert = new ClayAlert({
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			title: 'success',
		});

		alert.element.querySelector('button').click();

		expect(alert.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an alert after 5 seconds when type is "notification" and message has no links', function() {
		jest.useFakeTimers();

		alert = new ClayAlert({
			autoClose: true,
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			title: 'success',
			type: 'notification',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(5000);

		jest.runAllTimers();
		expect(alert.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an alert after 10 seconds when type is "notification" and message has links', function() {
		jest.useFakeTimers();

		alert = new ClayAlert({
			autoClose: true,
			closeable: true,
			message: 'message<a href="#">Link</a>',
			spritemap: spritemap,
			title: 'success',
			type: 'notification',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(10000);

		jest.runAllTimers();
		expect(alert.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an alert after 5 seconds when type is "stripe" and message has no links', function() {
		jest.useFakeTimers();

		alert = new ClayAlert({
			autoClose: true,
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			title: 'success',
			type: 'stripe',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(5000);

		jest.runAllTimers();
		expect(alert.visible_).toBeFalsy();
	});

	// eslint-disable-next-line
	it('should close automatically an alert after 10 seconds when type is "stripe" and message has links', function() {
		jest.useFakeTimers();

		alert = new ClayAlert({
			autoClose: true,
			closeable: true,
			message: 'message<a href="#">Link</a>',
			spritemap: spritemap,
			title: 'success',
			type: 'stripe',
		});

		expect(setTimeout.mock.calls.length).toBe(1);
		expect(setTimeout.mock.calls[0][1]).toBe(10000);

		jest.runAllTimers();
		expect(alert.visible_).toBeFalsy();
	});

	it('should close and destroy alert', function() {
		alert = new ClayAlert({
			closeable: true,
			destroyOnHide: true,
			message: 'message',
			spritemap: spritemap,
			title: 'success',
		});

		alert.element.querySelector('button').click();

		expect(alert.disposed_).toBeTruthy();
	});

	it('should fail when no title is passed', function() {
		expect(() => {
			alert = new ClayAlert({
				message: 'message',
				spritemap: spritemap,
			});
		}).toThrow();
	});

	it('should fail when no message is passed', function() {
		expect(() => {
			alert = new ClayAlert({
				spritemap: spritemap,
				title: 'title',
			});
		}).toThrow();
	});

	it('should fail when no spritemap is passed', function() {
		expect(() => {
			alert = new ClayAlert({
				message: 'message',
				title: 'title',
			});
		}).toThrow();
	});
});
