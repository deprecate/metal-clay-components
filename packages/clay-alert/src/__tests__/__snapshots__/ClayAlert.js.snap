// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayAlert should render a closeable alert 1`] = `
<div class="alert alert-info fade alert-dismissible show" role="alert">
  <button class="btn close btn-unstyled" aria-label="Close" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
    </svg>
  </button>
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">success</strong>message</div>
`;

exports[`ClayAlert should render a stripe alert 1`] = `
<div class="alert alert-info fade alert-fluid show" role="alert">
  <div class="container">
    <button class="btn close btn-unstyled" aria-label="Close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
      </svg>
    </button>
    <span class="alert-indicator">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
        <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
      </svg>
    </span>
    <strong class="lead">info</strong>message</div>
</div>
`;

exports[`ClayAlert should render a toast alert 1`] = `
<div class="alert alert-info fade alert-toast show" role="alert">
  <button class="btn close btn-unstyled" aria-label="Close" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#times"></use>
    </svg>
  </button>
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;

exports[`ClayAlert should render an alert style "danger" 1`] = `
<div class="alert alert-danger fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-exclamation-full">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#exclamation-full"></use>
    </svg>
  </span>
  <strong class="lead">danger</strong>message</div>
`;

exports[`ClayAlert should render an alert style "info" 1`] = `
<div class="alert alert-info fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;

exports[`ClayAlert should render an alert style "success" 1`] = `
<div class="alert alert-success fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-check-circle-full">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#check-circle-full"></use>
    </svg>
  </span>
  <strong class="lead">success</strong>message</div>
`;

exports[`ClayAlert should render an alert style "warning" 1`] = `
<div class="alert alert-warning fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-warning-full">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#warning-full"></use>
    </svg>
  </span>
  <strong class="lead">warning</strong>message</div>
`;

exports[`ClayAlert should render default markup 1`] = `
<div class="alert alert-info fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle">
      <use xlink:href="../node_modules/lexicon-ux/build/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;
